python/src/main/java/org/apache/apex/malhar/python/base/BasePythonExecutionOperator.java

```diff
@@ -0,0 +1,591 @@
+   * Represents all python commands that need to be run on a new instance of the operator after dynamic partitioning
+   */
+  private List<PythonRequestResponse> accumulatedCommandHistory = new ArrayList<>();
+
+  /***
+   * Processes the incoming tuple using the python engine that is injected. Also emits stragglers if any.
+   */
+  @InputPortFieldAnnotation
+  public final transient DefaultInputPort<T> input = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      numberOfRequestsProcessedPerCheckpoint += 1;
+      requestCounterForThisWindow += 1;
+      emitStragglers();
+      try {
+        PythonRequestResponse result = processPython(tuple,getApexPythonEngine());
+        if ( result != null) {
+          responseCounterForThisWindow += 1;
+          outputPort.emit(result);
+        } else {
+          numNullResponsesPerWindow += 1;
+        }
+      } catch (ApexPythonInterpreterException e) {
+        responseCounterForThisWindow += 1; // Error tuples need to be accounted for totals
+        errorPort.emit(tuple);
+        LOG.error("Error while processing tuple", e);
+      }
+    }
+  };
+
+
+  @Override
+  public void activate(Context.OperatorContext context)
+  {
+    getApexPythonEngine().setNumStarvedReturns(0L);
+  }
+
+  @Override
+  public void deactivate()
+  {
+
+  }
+
+  /***
+   * Used to emit the stragglers into the Stragglers port. INvoked both when a new tuple arrives or when Idle time
+   *  is detected on this operator.
+   */
+  private void emitStragglers()
+  {
+    List<PythonRequestResponse> stragglerResponse = new ArrayList<>();
+    getApexPythonEngine().getDelayedResponseQueue().drainTo(stragglerResponse);
+    for (PythonRequestResponse aReqResponse : stragglerResponse) {
+      responseCounterForThisWindow += 1;
+      stragglersPort.emit(aReqResponse);
+    }
+  }
+
+  /***
+   * Instantiates the configured python engine. Only in-memory implementation provided for now
+   * @param context The operator context
+   * @return The python engine
+   */
+  protected ApexPythonEngine initApexPythonEngineImpl(Context.OperatorContext context)
+  {
+    JepPythonEngine jepPythonEngine = new JepPythonEngine("" + context.getId(),workerThreadPoolSize);
+    jepPythonEngine.setSleepTimeAfterInterpreterStart(sleepTimeDuringInterpreterBoot);
+    return jepPythonEngine;
+  }
+
+  /***
+   * Instantiates the partitioner. Only Thread Starvation based partitioner for now.
+   */
+  private void initPartitioner()
+  {
+    if (partitioner == null) {
+      synchronized (objectForLocking) {
+        if (partitioner == null) {
+          switch (partitionerType) {
+            default:
+            case THREAD_STARVATION_BASED:
+              partitioner = new ThreadStarvationBasedPartitioner(this);
+              break;
+          }
+        }
+      }
+    }
+  }
+
+  /***
+   * Starts the python engine and also sleeps for sometime ( configurable) to ensure that the interpreter is
```

Comment:
sleep intervals are always problematic, they result in non-deterministic behavior

Reply:
The sleep can be configured to be 0 in which case the thread will continue ahead. It was only in the case of Tensorflow testing that this sleep was required because the complete loading of the tensor flow library seem to have been done on another thread. In this case, immediately issuing a python command resulted in errors. The only way seemed to be letting the main thread to sleep. A busy spin might not solve the problem either. The other plausible way is to use a test command that does not interfere with the users interpreter commands ( and the command makes use of the library tensorflow in this case ) and the main thread continuously issues this configured command and see if the return status is true to signify the completion of the boot up of the interpreter.